systems/plants/collision/BulletModel.cpp

```diff
@@ -23,113 +23,329 @@ namespace DrakeCollision
+          }
+          break;
+        case DrakeShapes::CAPSULE:
+          {
+            const auto capsule = static_cast<const DrakeShapes::Capsule&>(elements[id]->getGeometry());
+            bt_shape = newBulletCapsuleShape(capsule, true);
+            bt_shape_no_margin = newBulletCapsuleShape(capsule, false);
+          }
+          break;
+        default:
+          cerr << "Warning: Collision elements[id] has an unknown type " << elements[id]->getShape() << endl;
+          throw unknownShapeException(elements[id]->getShape());
+          break;
+      }
+      if (bt_shape) {
+        // Create the collision objects
+        unique_ptr<btCollisionObject> bt_obj(new btCollisionObject());
+        unique_ptr<btCollisionObject> bt_obj_no_margin(new btCollisionObject());
+        bt_obj->setCollisionShape(bt_shape.get());
+        bt_obj_no_margin->setCollisionShape(bt_shape_no_margin.get());
+        bt_obj->setUserPointer(elements[id].get());
+        bt_obj_no_margin->setUserPointer(elements[id].get());
+
+        // Add the collision objects to the collision worlds
+        bullet_world.bt_collision_world->addCollisionObject(bt_obj.get());
+        bullet_world_no_margin.bt_collision_world->addCollisionObject(bt_obj_no_margin.get());
+
+        if (elements[id]->isStatic()) {
+          bt_obj->setCollisionFlags(btCollisionObject::CF_KINEMATIC_OBJECT);
+          bt_obj->activate();
+          bt_obj_no_margin->setCollisionFlags(btCollisionObject::CF_KINEMATIC_OBJECT);
+          bt_obj_no_margin->activate();
+        } else {
+          bt_obj->setCollisionFlags(btCollisionObject::CF_STATIC_OBJECT);
+          bt_obj_no_margin->setCollisionFlags(btCollisionObject::CF_STATIC_OBJECT);
+        }
+
+        // Store the Bullet collision objects
+        bullet_world.bt_collision_objects.insert(make_pair(id, move(bt_obj)));
+        bullet_world_no_margin.bt_collision_objects.insert(make_pair(id, move(bt_obj_no_margin)));
+
+        // Store the Bullet collision shapes too, because Bullet does no cleanup
+        bt_collision_shapes.push_back(move(bt_shape));
+        bt_collision_shapes.push_back(move(bt_shape_no_margin));
       }
-      return true;
-    } else {
-      return false;
     }
+    return id;
   }
 
-  bool BulletModel::setCollisionFilter(const int body_idx,
-                                      const uint16_t group, 
-                                      const uint16_t mask)
+  ElementId BulletModel::addElement(std::unique_ptr<Element> element)
   {
-    auto iter_for_body( bodies.find(body_idx) );
-    if ( iter_for_body!=bodies.end() ) { // Then bodies[body_idx] exists
-      setCollisionFilter(iter_for_body->second,group,mask);
-      return true;
-    } else {
-      return false;
+    ElementId id =  Model::addElement(move(element));
+
+    if (id != 0) {
+      unique_ptr<btCollisionShape> bt_shape;
+      unique_ptr<btCollisionShape> bt_shape_no_margin;
+      switch (elements[id]->getShape()) {
```

Comment:
this looks like a lot of code duplication from the lines above.  can't you just have one addElement call the other?